src/client/grid/hex.tsx

```diff
@@ -118,20 +117,51 @@ export function Hex({ space, selectedGood, highlightedTrack, size, hideGoods, cl
-    {gameKey === 'cyprus' && <CyprusBorder space={space} center={center} size={size} />}
-    {trackInfo.map((t, index) => <TrackSvg key={index} center={center} size={size} track={t} highlighted={highlightedTrackSet.has(t)} rotation={rotation} />)}
-    {space instanceof Land && (space.getTileType() != null ? isTownTile(space.getTileType()!) : space.hasTown()) && <circle cx={center.x} cy={center.y} fill="white" r={size / 2} />}
-    {space instanceof Land && space.getTileType() == null && space.getTerrainCost() != null && <TerrainCost space={space} center={center} size={size} rotation={rotation} />}
-    {space instanceof Land && space.hasTown() && <HexName name={space.name()!} rotation={rotation} center={center} size={size} />}
-    {space instanceof City && space.onRoll().length > 0 && <OnRoll city={space} center={center} size={size} rotation={rotation} />}
-    {space instanceof City && space.name() != '' && <HexName name={space.name()} rotation={rotation} center={center} size={size} isCity={true} />}
-    {space instanceof City && !hideGoods && space.getGoods().map((g, index) => <GoodBlock key={index} clickable={clickTargets.has(ClickTarget.GOOD)} coordinates={coordinates} highlighted={selectedGoodIndex === index} offset={index} good={g} center={center} size={size} rotation={rotation} />)}
-  </>;
+  if (space instanceof Land) {
+    const hexColor = landColorStyle(space);
+
+    return <>
+      <polygon className={`${styles.location} ${clickable ? gridStyles.clickable : ''} ${hexColor}`} data-coordinates={space.coordinates.serialize()} points={corners} stroke="black" strokeWidth="0" />
+      <polygon fillOpacity="0" data-coordinates={space.coordinates.toString()} points={corners} stroke="black" strokeWidth={size / 100} />
+      {space.unpassableExits().map(direction => <EdgeBoundary key={direction} center={center} size={size} direction={direction} />)}
+      {gameKey === 'cyprus' && <CyprusBorder space={space} center={center} size={size} />}
+      {trackInfo.map((t, index) => <TrackSvg key={index} center={center} size={size} track={t} highlighted={highlightedTrackSet.has(t)} rotation={rotation} />)}
+      {(space.getTileType() != null ? isTownTile(space.getTileType()!) : space.hasTown()) && <circle cx={center.x} cy={center.y} fill="white" r={size * 0.4} />}
+      {space.getTileType() == null && space.getTerrainCost() != null && <TerrainCost space={space} center={center} size={size} rotation={rotation} />}
+      {space.hasTown() && <HexName name={space.name()!} rotation={rotation} center={center} size={size} />}
+    </>;
+  } else {
+    const [hexColor, alternateColor] = cityColorStyles(space);
+
+    const onRoll = space.onRoll();
+    let cityGroup: CityGroup;
+    if (onRoll.length > 0) {
+      cityGroup = onRoll[0].group;
+    } else {
+      cityGroup = CityGroup.WHITE;
+    }
+
+    let outerFill = '#ffffff';
+    if (cityGroup == CityGroup.BLACK) {
+      outerFill = '#222222';
```

Comment:
The logic for cityGroup/outerFill isn't that intuitive and it may get lost as things change. I'd recommend just creating a `const outerFill = calcOuterFill();` then burying this logic there.

Or you could use `useMemo` as an easy way to contain the logic.

```
const onRoll = space.onRoll()[0]?.group;
const outerFill = useMemo(() => {
  if (onRoll == null || onRoll === CityGroup.WHITE) return '#ffffff';
  return '#222222';
}, [onRoll]);